include/openssl/ssl.h

```diff
@@ -1949,6 +1950,18 @@ void SSL_set_not_resumable_session_callback(SSL *ssl,
                                             int (*cb) (SSL *ssl,
                                                        int
                                                        is_forward_secure));
+
+void SSL_CTX_set_choose_cipher_callback(SSL_CTX *ctx,
+                                        const SSL_CIPHER* (*cb) (SSL *ssl,
+                                                                 STACK_OF(SSL_CIPHER) *client));
+
+void SSL_set_choose_cipher_callback(SSL *ssl,
+                                    const SSL_CIPHER* (*cb) (SSL *ssl,
+                                                             STACK_OF(SSL_CIPHER) *client));
+
+
+const SSL_CIPHER *SSL_choose_cipher_default(SSL *s, STACK_OF(SSL_CIPHER) *clnt);
```

Comment:
One possibility is to move the callback calling code into ssl3_choose_cipher() itself. If the callback returns NULL then just continue with the default logic. That would exclude the possibility of the callback deciding that no ciphersuites are acceptable of course - we'd have to do something different if that is a requirement. It would also make detecting errors occurring in the callback more difficult.

Perhaps add a const SSL_CIPHER ** argument to the callback with an int return code, e.g. 0 = error, 1 = use callback provided cipher, 2 = continue with default logic. I'd rather do that then expose an additional function in the API.